plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/PurityChecker.kt

```diff
@@ -0,0 +1,60 @@
+package org.jetbrains.kotlin.formver.purity
+
+import org.jetbrains.kotlin.formver.embeddings.callables.UserFunctionEmbedding
+import org.jetbrains.kotlin.formver.embeddings.expression.*
+
+class PurityChecker() {
+
+    // Keeping these because they might be useful later?
+    /*
+    private var debugSink: ((String) -> Unit)? = null
+    fun setDebugSink(sink: ((String) -> Unit)?) {
+        debugSink = sink
+    }
+
+    private inline fun dbg(msg: () -> String) {
+        debugSink?.invoke(msg())
+    }
+    */
+
+    /**
+     * We restrict our selves to using a helper function for now that only evaluates asserts.
+     */
+    fun checkIsPure(userFunction: UserFunctionEmbedding) {
+        val exp = userFunction.body ?: return
+        assertsArePure(exp.debugExpEmbedding!!)
+    }
```

Comment:
Typically there are two ways of traversing a tree defined through a class hierarchy:
1. Add a virtual function that does the work.
2. Define a visitor type and implement the traversal as a visitor.

I think either one is possible here; I suspect a `checkValidity` virtual function that just calls itself on all children + calls `checkOwnValidity` that actually does work (for nodes where it's relevant, in your case just `Assert`) would be the simplest solution.